src/main/java/com/fauna/codec/codecs/DynamicCodec.java

```diff
@@ -0,0 +1,218 @@
+package com.fauna.codec.codecs;
+
+import com.fauna.codec.Codec;
+import com.fauna.codec.CodecProvider;
+import com.fauna.enums.FaunaTokenType;
+import com.fauna.exception.ClientException;
+import com.fauna.exception.NullDocumentException;
+import com.fauna.serialization.*;
+import com.fauna.types.*;
+import com.fauna.types.Module;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DynamicCodec extends BaseCodec<Object> {
+    private final ListCodec<Object, List<Object>> list = new ListCodec<>(this);
+    private final PageCodec<Object, Page<Object>> page = new PageCodec<>(this);
+    private final MapCodec<Object, Map<String, Object>> map = new MapCodec<>(this);
+    private final CodecProvider provider;
+
+    public DynamicCodec(CodecProvider provider) {
+
+        this.provider = provider;
+    }
+
+    @Override
+    public Object decode(UTF8FaunaParser parser) throws IOException {
+        Object value = null;
+        switch (parser.getCurrentTokenType()) {
+            case NULL:
+                break;
+            case START_OBJECT:
```

Comment:
A strategy I've sometimes used because Java is dumb and makes you have the break statement is to return directly from the switch statement. It seems cleaner to me, but again, just an opinion.